### PR TITLE
Add Earliest Deadline First scheduling code

### PR DESCRIPTION
**Pull request by KelvinPuyam**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by KelvinPuyam** [comment]
Requesting review: @cclauss

File: scheduling/earliest_deadline_first.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    Calculate the average waiting time.
+    Return: The average waiting time.
+    """
+    return sum(waiting_times) / len(waiting_times)
+
+
+if __name__ == "__main__":
+    arrival_times = [0, 1, 2]
+    burst_times = [3, 5, 2]
+    deadlines = [5, 7, 6]
```
Comment (cclauss) [inline scheduling/earliest_deadline_first.py:77]:
```suggestion
from typing import NamedTuple


class Job(NamedTuple):
    arrival_time: int
    burst_time: int
    deadline: int
    __slots__ = ()

    jobs = (Job(0, 3, 5), Job(1, 5, 7), Job(2, 2, 6))
```

File: scheduling/earliest_deadline_first.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+Earliest Deadline First (EDF) scheduling algorithm
+In EDF, the process with the earliest deadline is selected for execution.
+https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling
+Author: KelvinPuyam
+"""
+
+from __future__ import annotations
+
+
+def calculate_waiting_times(
```
Comment (cclauss) [inline scheduling/earliest_deadline_first.py:11]:
```suggestion
from typing import NamedTuple


class Job(NamedTuple):
    arrival_time: int
    burst_time: int
    deadline: int
    __slots__ = ()


def calculate_waiting_times(
```

**Comment by cclauss** [comment]
The best way to prove that you are doing this right would be to take one or both of the diagrams in the Wikipedia article and ensure that your results exactly match their results.  If instance...
* https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling#/media/File:EDF_Scheduling_test_failed.png

Use the same Job field names.  Perhaps you could use DataClass instead of a NamedTuple so you can add `wait_time` to each job, etc.

**Pull request closed without merging**